src/System.Private.ServiceModel/src/System/ServiceModel/Channels/HttpResponseMessageHelper.cs

```diff
@@ -0,0 +1,367 @@
+
+            return message;
+        }
+
+        private Exception ProcessHttpAddressing(Message message)
+        {
+            Exception result = null;
+            AddProperties(message);
+
+            // check if user is receiving WS-1 messages
+            if (message.Version.Addressing == AddressingVersion.None)
+            {
+                bool actionAbsent = false;
+                try
+                {
+                    actionAbsent = (message.Headers.Action == null);
+                }
+                catch (XmlException)
+                {
+                }
+                catch (CommunicationException)
+                {
+                }
+                if (!actionAbsent)
+                {
+                    result = new ProtocolException(SR.Format(SR.HttpAddressingNoneHeaderOnWire,
+                        XD.AddressingDictionary.Action.Value));
+                }
+
+                bool toAbsent = false;
+                try
+                {
+                    toAbsent = (message.Headers.To == null);
+                }
+                catch (XmlException)
+                {
+                }
+                catch (CommunicationException)
+                {
+                }
+
+                if (!toAbsent)
+                {
+                    result = new ProtocolException(SR.Format(SR.HttpAddressingNoneHeaderOnWire,
+                        XD.AddressingDictionary.To.Value));
+                }
+                message.Headers.To = message.Properties.Via;
+            }
+
+            return result;
+        }
+
+        private void AddProperties(Message message)
+        {
+            HttpResponseMessageProperty responseProperty = new HttpResponseMessageProperty(_httpResponseMessage);
+            message.Properties.Add(HttpResponseMessageProperty.Name, responseProperty);
+            message.Properties.Via = message.Version.Addressing.AnonymousUri;
+        }
+
+        private async Task<bool> ValidateContentTypeAsync()
+        {
+            var content = _httpResponseMessage.Content;
+            if (content != null)
+            {
+                var mediaValueContentType = content.Headers.ContentType;
+                _contentType = mediaValueContentType == null ? string.Empty : mediaValueContentType.ToString();
+                _contentLength = content.Headers.ContentLength.HasValue ? content.Headers.ContentLength.Value : -1;
+            }
+
+            if(string.IsNullOrEmpty(_contentType))
+            {
+                Stream contentStream = await GetStreamAsync();
+                if (contentStream != null)
+                {
+                    throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(new ProtocolException(SR.HttpContentTypeHeaderRequired));
+                }
+                return false;
+            }
+            else if(_contentLength!=0)
+            {
```

Comment:
_contentLength and _contentType already have the correct default values (0 and null) in the case that content==null. _contentLength==-1 means the content length wasn't specified, so presume chunked encoding.